### ISSUE TITLE
Type unions not supported

### ISSUE THREAD

**Issue by ArneBachmannDLR** [mkdocstrings/mkdocstrings#314]
**Describe the bug**
My code uses *type unions* like `a:str|int`, or even `var:str|Tuple[int,int]` (which combines concrete with generic types).
This is usually made backwards-compatible for Python < 3.10 via `from __future__ import annotations`.
When using mkdocstrings via `mkdocs build`, however, the code fails in several ways, unless using the classic `Union[int,str]` explicitly (which I don't want).

**To Reproduce**
With `from __future__` as first import:
- `TypeError: typing.Final[logging.Logger] is not valid as type argument` - this error message is wrong and misleading, because the argument (Logger) is NOT a generic.

Without `from __future__`:
- `TypeError: unsupported operand type(s) for |: 'type' and '_GenericAlias'` - OK, now we see, this is a type union problem, which doesn't seem to be supported on my Python 3.8 (hence the use of `annotations`).

**Expected behavior**
`mkdocs build` should run without problems even on Python 3.8, just like it does when using `Union[int,str]` instead of `int|str`.

**Information**
- OS: Windows 20H2
- Anaconda Python 3.8.8 with pip-installed mkdocs, mkdocstrings
- `mkdocstrings` version: 0.15.2

**Stack trace**
```
           Traceback (most recent call last):
              File "d:\apps\miniforge3\lib\site-packages\pytkdocs\cli.py", line 205, in main
                output = json.dumps(process_json(line))
              File "d:\apps\miniforge3\lib\site-packages\pytkdocs\cli.py", line 114, in process_json
                return process_config(json.loads(json_input))
              File "d:\apps\miniforge3\lib\site-packages\pytkdocs\cli.py", line 91, in process_config
                obj = loader.get_object_documentation(path, members)
              File "d:\apps\miniforge3\lib\site-packages\pytkdocs\loader.py", line 353, in get_object_documentation
                leaf = get_object_tree(dotted_path, self.new_path_syntax)
              File "d:\apps\miniforge3\lib\site-packages\pytkdocs\loader.py", line 252, in get_object_tree
                parent_module = importlib.import_module(parent_module_path)
              File "d:\apps\miniforge3\lib\importlib\__init__.py", line 127, in import_module
                return _bootstrap._gcd_import(name[level:], package, level)
              File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
              File "<frozen importlib._bootstrap>", line 991, in _find_and_load
              File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
              File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
              File "<frozen importlib._bootstrap_external>", line 783, in exec_module
              File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
              File "D:\forks\HACE\logfiler\logfiler\api.py", line 32, in <module>
                class OutputCallback(Protocol):
              File "D:\forks\HACE\logfiler\logfiler\api.py", line 33, in OutputCallback
                def __call__(_, s:str|Tuple[str,int,str], *args, **kwds) -> None: ...
            TypeError: unsupported operand type(s) for |: 'type' and '_GenericAlias'
```

and with `__future__`

```
            Traceback (most recent call last):
              File "d:\apps\miniforge3\lib\site-packages\pytkdocs\cli.py", line 205, in main
                output = json.dumps(process_json(line))
              File "d:\apps\miniforge3\lib\site-packages\pytkdocs\cli.py", line 114, in process_json
                return process_config(json.loads(json_input))
              File "d:\apps\miniforge3\lib\site-packages\pytkdocs\cli.py", line 91, in process_config
                obj = loader.get_object_documentation(path, members)
              File "d:\apps\miniforge3\lib\site-packages\pytkdocs\loader.py", line 356, in get_object_documentation
                root_object = self.get_module_documentation(leaf, members)
              File "d:\apps\miniforge3\lib\site-packages\pytkdocs\loader.py", line 416, in get_module_documentation
                attributes_data = get_module_attributes(module)
              File "d:\apps\miniforge3\lib\site-packages\pytkdocs\parsers\attributes.py", line 106, in get_module_attributes
                return combine(get_module_or_class_attributes(get_nodes(module)), get_type_hints(module))
              File "d:\apps\miniforge3\lib\typing.py", line 1264, in get_type_hints
                value = _eval_type(value, globalns, localns)
              File "d:\apps\miniforge3\lib\typing.py", line 270, in _eval_type
                return t._evaluate(globalns, localns)
              File "d:\apps\miniforge3\lib\typing.py", line 517, in _evaluate
                self.__forward_value__ = _type_check(
              File "d:\apps\miniforge3\lib\typing.py", line 142, in _type_check
                raise TypeError(f"{arg} is not valid as type argument")
            TypeError: typing.Final[logging.Logger] is not valid as type argument
```

**Comment by ArneBachmannDLR** [comment]
Hint: When running with Python 3.10 (alpha), I get something else entirely:

```
            Traceback (most recent call last):
              File "d:\apps\py310\lib\site-packages\pytkdocs\cli.py", line 205, in main
                output = json.dumps(process_json(line))
              File "d:\apps\py310\lib\site-packages\pytkdocs\cli.py", line 114, in process_json
                return process_config(json.loads(json_input))
              File "d:\apps\py310\lib\site-packages\pytkdocs\cli.py", line 91, in process_config
                obj = loader.get_object_documentation(path, members)
              File "d:\apps\py310\lib\site-packages\pytkdocs\loader.py", line 353, in get_object_documentation
                leaf = get_object_tree(dotted_path, self.new_path_syntax)
              File "d:\apps\py310\lib\site-packages\pytkdocs\loader.py", line 269, in get_object_tree
                obj = getattr(current_node.obj, obj_name)
            AttributeError: module 'logfiler' has no attribute 'api'
```

so my package `logfiler` with module (aka file) `api.py` is not even found.

**Comment by pawamoy** [comment]
The limitations of the current pytkdocs implementation start to really show up :sweat_smile: 
Indeed, mkdocstrings/pytkdocs must be run using the same Python version your code is compatible with. With 3.10 features in your code, your must run mkdocstrings/pytkdocs with 3.10 as well (unless the features are made backward compatible in some ways).

Next thing is pytkdocs does not support concrete type annotations yet.

The `module 'logfiler' has no attribute 'api'` might be because of something else though. Do you have a public repo I could use to reproduce?

Also note that I'm currently working on a new implementation of pytkdocs that would lift up all these limitations.

**Comment by ArneBachmannDLR** [comment]
Thanks for the quick reply!
I managed to get around the import error - under Python 3.10 I was missing a dependency which was imported from another package, which then missed the api module.

The error, however, persists even with Python 3.10b1:
`TypeError: typing.Final[logging.Logger] is not valid as type argument`, so I guess this is what you meant by `does not support concrete type annotations yet.` - while most type annotation *do* work, just not type unions via `|` or the more recent additions like `Final`, `Literal` or `Protocol`. I import them like that:
```
try:    from typing            import Final  # 3.8+
except: from typing_extensions import Final  # type: ignore
```

**Comment by pawamoy** [comment]
What I meant is that in some specific cases, pytkdocs uses `ast` to parse annotations, and it probably wouldn't understand `|` or `list[dict[str, int]]` and similar annotations. But in your case it seems Python itself is throwing a TypeError. Not sure what's happening here.

**Comment by ArneBachmannDLR** [comment]
Ok, thanks so far. I think I cannot further assist you in this issue's resolution, as this code base is totally unknown to me.

**Comment by pawamoy** [comment]
Turns out the solution is pretty simple:

```diff
# src/pytkdocs/loader.py

+ from __future__ import annotations
```

We would also need to drop support for Python 3.6, as this future is not available for it.

**Comment by ArneBachmannDLR** [comment]
I tried it - there is no complaint about the union pipe anymore, which is great!

But I'm still stuck with the next error, which I described in my original post:
```
TypeError: typing.Final[str] is not valid as type argument
```

Removing the `str` leads to:
```
 TypeError: Plain typing.Final is not valid as type argument
```

This issue is probably a deeper problem when interacting with typing, I don't know.
Should we split a different ticket for that?

**Comment by ArneBachmannDLR** [comment]
BTW the error goes away if I remove the `from __future__`in both `loader.py` AND my source file.

**Comment by pawamoy** [comment]
Could you provide a small Python file with the related code? I can't replicate the `Final` issue. Could it be an issue with miniconda?

**Comment by pawamoy** [comment]
Could even be a bug in the standard `typing` module.

**Comment by pawamoy** [comment]
I opened https://bugs.python.org/issue45166

**Comment by ArneBachmann** [comment]
Thanks for taking care of this!

**Comment by pawamoy** [comment]
The fix was merged in 3.11 and backported to 3.10 and 3.9 (not 3.8?), so I'll close this in favor of https://github.com/mkdocstrings/pytkdocs/pull/113
